### ISSUE TITLE
Update implementation based on new APIs

### ISSUE THREAD

**Issue by gr2m** [gr2m/github-project#56]
### Please avoid duplicates

- [X] I checked [all open feature requests](https://github.com/gr2m/github-project/issues?q=is%3Aissue+is%3Aopen+label%3Afeature) and none of them matched my problem.

### What’s missing?

There are lots of new GraphQL query & mutation nodes that allow for more efficient implementations, and now also support creating draft issues: https://github.blog/changelog/2022-06-23-the-new-github-issues-june-23rd-update/

### Why?

For many operations, we currently download all of the project items, which can be up to 1200! As an example, with the new APIs, items can be accessed through a pull request / issue node, so we don't need to download all items and filter them by their item contents afterwards, we can access the item directly based on repository & issue/PR information.

### Alternatives you tried

n/a

### Would you be interested in contributing the feature?

- [ ] yes

**Comment by gr2m** [comment]
The most important optimization is that we no longer need to load all items in order to retrieve an item based on content node ID or repository name + issue number, as there is now the `projectItems` node on the `Issue` and `PullRequest` type

Something like this, I'm not sure if there is a better way to retrieve all the field values, I didn't look deeply into it.

```graphql
{
  repository(owner: "github", name: "roadmap") {
    issue(number: 195) {
      projectItems(first: 1) {
        nodes {
          fieldValues(first: 20) {
            nodes {
              __typename
              ... on ProjectV2ItemFieldDateValue {
                date
                field {
                  ... on ProjectV2Field {
                    id
                    name
                  }
                }
              }
              ... on ProjectV2ItemFieldIterationValue {
                title
                iterationId
                startDate
                duration
                field {
                  ... on ProjectV2Field {
                    id
                    name
                  }
                }
              }
              ... on ProjectV2ItemFieldNumberValue {
                number
                field {
                  ... on ProjectV2Field {
                    id
                    name
                  }
                }
              }
              ... on ProjectV2ItemFieldSingleSelectValue {
                optionId
                name
                field {
                  ... on ProjectV2Field {
                    id
                    name
                  }
                }
              }
              ... on ProjectV2ItemFieldTextValue {
                text
                field {
                  ... on ProjectV2Field {
                    id
                    name
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
```

The mutation to create a draft item would look something like this

```js
mutation addProjectV2DraftIssue($projectId: ID!, $title: String!, $body: String, $assigneeIds: [ID!]) {
  addProjectV2DraftIssue(input: {projectId: $projectId, title: $title, body: $body, assigneeIds: $assigneeIds}) {
    projectItem {
      id
      content {
        ... on DraftIssue {
          id
          title
          body
        }
      }
    }
  }
}
```

I also found that all the node IDs changed between the `ProjectNext` and `ProjectV2`, so we should release this as a breaking change in case users persist IDs in some way, they would need to migrate their data.

**Comment by gr2m** [comment]
Query to retrieve project core data

```graphql
query getProjectCoreData($org: String!, $projectNumber: Int!) {
  organization(login: $org) {
    projectV2(number: $projectNumber) {
      id
      title
      readme
      url
      fields(first: 20) {
        nodes {
          ... on ProjectV2FieldCommon {
            id
            dataType
            name
          }
          ... on ProjectV2SingleSelectField {
            options {
              id
              name
            }
          }
          ... on ProjectV2IterationField {
            configuration {
              iterations {
               	title
                duration
                startDate
              }
              completedIterations {
               	title
                duration
                startDate
              }
              duration
              startDay
            }
          }
        }
      }
      items(first: 100) {
        nodes {
          id
          createdAt
          content {
            __typename
            ... on DraftIssue {
              id
              title
              createdAt
              updatedAt
              author: creator {
                login
              }
              assignees(first: 10) {
                nodes {
                  login
                }
              }
            }
            ... on Issue {
              id
              databaseId
              number
              title
              createdAt
              author {
                login
              }
              assignees(first: 10) {
                nodes {
                  login
                }
              }
              labels(first: 10) {
                nodes {
                  name
                }
              }
              closed
              milestone {
                number
                title
                state
              }
              repository {
                nameWithOwner
              }
            }
            ... on PullRequest {
              id
              databaseId
              number
              title
              createdAt
              author {
                login
              }
              assignees(first: 10) {
                nodes {
                  login
                }
              }
              labels(first: 10) {
                nodes {
                  name
                }
              }
              closed
              milestone {
                number
                title
                state
              }
              repository {
                nameWithOwner
              }
              merged
            }
          }
          fieldValues(first: 20) {
            nodes {
              __typename
              ... on ProjectV2ItemFieldDateValue {
                date
                field {
                  ... on ProjectV2Field {
                    id
                  }
                }
              }
              ... on ProjectV2ItemFieldIterationValue {
                title
                iterationId
                startDate
                duration
                field {
                  ... on ProjectV2IterationField {
                    id
                  }
                }
              }
              ... on ProjectV2ItemFieldNumberValue {
                number
                field {
                  ... on ProjectV2Field {
                    id
                  }
                }
              }
              ... on ProjectV2ItemFieldSingleSelectValue {
                optionId
                field {
                  ... on ProjectV2SingleSelectField {
                    id
                  }
                }
              }
              ... on ProjectV2ItemFieldTextValue {
                text
                field {
                  ... on ProjectV2Field {
                    id
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
```

**Comment by gr2m** [comment]
I made some progress, see the [`56-use-project-v2-nodes` branch](https://github.com/gr2m/github-project/tree/56-use-project-v2-nodes).

One thing I realized is that draft items have content, at least when using the ProjectV2  nodes. I can see the following keys:

- `id`
- `title`
- `createdAt`
- `updatedAt`
- `author`
- `assignees`

I'll add these to the changes I'm working on.

Other than that, I still need to create the new API to create a draft item, and figure out how to create a maintainable test for testing the use case when an actor has access to a project, but not to an item in it (because the item content is in a private repository that the actor has no access to). Might not be worth the hassle but I'll try
